### ISSUE TITLE
#[main] attribute was unintentionally stabilized (as no-op) in 1.53.0

### ISSUE THREAD

**Issue by jeremyBanks** [rust-lang/rust#93786]
### Code

I tried this code:

```rust
#[main]
fn main() {}
```

I expected an error, because the main attribute has never been available on stable, and had apparently been removed in #84217.

Instead, the program ran without error, although the attribute has no apparent effect.

### Version it worked on

It most recently worked as expected (emitting an error) on Rust 1.52.1:

```
$ cargo +1.52.1 build    
   Compiling regression-testing v0.0.0 (/Users/jeremy/src/github.com/jeremyBanks/rust-regression)
error[E0658]: declaration of a non-standard `#[main]` function may change over time, for now a top-level `fn main()` is required
 --> src/main.rs:2:1
  |
2 | fn main() {}
  | ^^^^^^^^^^^^
  |
  = note: see issue #29634 <https://github.com/rust-lang/rust/issues/29634> for more information
```

### Version with regression

This stopped being an error in 1.53.0 (in which a warning was still emitted), and continues to be accepted in stable through the current 1.58.1 version (in which the warning has also disappeared).

```
$ cargo +1.53.0 build    
   Compiling regression-testing v0.0.0 (/Users/jeremy/src/github.com/jeremyBanks/rust-regression)
warning: unused attribute
 --> src/main.rs:1:1
  |
1 | #[main]
  | ^^^^^^^
  |
  = note: `#[warn(unused_attributes)]` on by default
```

```
$ cargo +1.58.1 build    
   Compiling regression-testing v0.0.0 (/Users/jeremy/src/github.com/jeremyBanks/rust-regression)
```

### Potential Solution

I submitted PR #93753 with the simplest potential solution: entirely removing the `#[main]` attribute. However, other potential solutions are being discussed in the comments so it seemed like this warranted an issue for the bug instead of just a PR for my specific solution.

---

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

**Comment by jeremyBanks** [comment]
Copying over @pnkfelix's https://github.com/rust-lang/rust/pull/93753#issuecomment-1032576085 from the PR:

> If I understand correctly, the main effect of this PR is to make code start erroring that is using the previously-built-in no-op `#[main]`. Any such code must inherently have that attribute by accident (since the attribute cannot be serving a purpose for any code that is using it).
> 
> I don't think I would beta backport this without at least a crater run, since we don't know how many crates out there might be using `#[main]` by accident.
> 
> It also looks like something that would be nearly trivial to put through a future-incompatibility warning cycle (which I know is essentially the opposite direction from a beta backport, but the nature of this problem seems like it _deserves_ that treatment: i.e. its not an urgent matter to fix this).
> 
> The only argument I can see _against_ a warning cycle is that its just extra tech debt hanging around in the compiler. Versus this PR, which immediately eliminates that tech debt. So this is sort of a microcosm of an ongoing prioritization question: should we prioritize the needs of an unknown but likely small set of Rust users over the needs of rustc contributors...

**Comment by apiraino** [comment]
Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2393786.20.23.5Bmain.5D.20attribute.20was.20accidentally.20stabilized.20.28as.20no.E2.80.A6) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-high +T-compiler
